### PR TITLE
WIP

### PR DESCRIPTION
**Pull request by geropl**
## Description

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at bdd76a5</samp>

Refactor and simplify permission checks on workspaces and projects in `workspace-service.ts` and add a helper function for missing instances.

</details>

## Related Issue(s)

Fixes #

#### Preview status

gitpod:summary

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/authorization/authorizer.ts [hunk 1]
```diff
@@ -116,40 +118,61 @@ export class Authorizer {
         );
     }
 
-    async hasPermissionOnProject(userId: string, permission: ProjectPermission, projectId: string): Promise<boolean> {
+    async hasPermissionOnProject(
+        userId: string,
+        permission: ProjectPermission,
+        project: { id: string; teamId?: string },
```
Comment (svenefftinge) [inline components/server/src/authorization/authorizer.ts:124]: lets make this explicit and pass in `projectId: string` and `organizationId: string`

File: components/server/src/authorization/authorizer.ts [hunk 2]
```diff
@@ -116,40 +118,61 @@ export class Authorizer {
...
+            `You do not have ${permission} on project ${project.id}`,
         );
     }
 
-    async hasPermissionOnUser(userId: string, permission: UserPermission, resourceUserId: string): Promise<boolean> {
+    async hasPermissionOnUser(
+        userId: string,
+        permission: UserPermission,
+        resourceUserId: string,
+        resourceUser: Pick<User, "organizationId"> | undefined,
```
Comment (svenefftinge) [inline components/server/src/authorization/authorizer.ts:167]: lets make this explicit and pass in `organizationId: string`

File: components/server/src/authorization/authorizer.ts [hunk 3]
```diff
@@ -178,34 +210,42 @@ export class Authorizer {
     async hasPermissionOnWorkspace(
         userId: string,
         permission: WorkspacePermission,
-        workspaceId: string,
+        workspace: { id: string; organizationId: string | undefined },
```
Comment (svenefftinge) [inline components/server/src/authorization/authorizer.ts:213]: same here -> lets make this explicit and pass in `workspaceId: string` and `organizationId: string`

File: components/server/src/authorization/authorizer.ts (hunk 2 above)
Comment (geropl) [reply components/server/src/authorization/authorizer.ts:167]: I'd much rather pass in `User`: it's much more intuitive, and less prone to errors. Also, we need to distinguish between "User with organizationId === undefined" and "no user passed (shape === undefined)". This way we keep the meaning + make it easy to use from the outside.

**Pull request closed without merging**